consensus/src/liveness/leader_reputation.rs

```diff
@@ -127,70 +125,57 @@ impl ActiveInactiveHeuristic {
             .filter_map(|(validator, &voted)| if voted { Some(validator) } else { None })
             .collect())
     }
-}
 
-#[cfg(test)]
-mod tests {
-    use crate::liveness::leader_reputation::ActiveInactiveHeuristic;
-    use aptos_types::account_address::AccountAddress;
-
-    #[test]
-    fn test_bitmap_to_voters() {
-        let validators: Vec<_> = (0..4)
-            .into_iter()
-            .map(|_| AccountAddress::random())
-            .collect();
-        let bitmap = vec![true, true, false, true];
-
-        if let Ok(voters) = ActiveInactiveHeuristic::bitmap_to_voters(&validators, &bitmap) {
-            assert_eq!(&validators[0], voters[0]);
-            assert_eq!(&validators[1], voters[1]);
-            assert_eq!(&validators[3], voters[2]);
-        }
+    pub fn indices_to_validators<'a>(
+        validators: &'a [Author],
+        indices: &[u64],
+    ) -> Result<Vec<&'a Author>, String> {
+        indices
+            .iter()
+            .map(|index| {
+                usize::try_from(*index)
+                    .map_err(|_err| format!("index {} out of bounds", index))
+                    .and_then(|index| {
+                        validators.get(index).ok_or(format!(
+                            "index {} is larger than number of validators {}",
+                            index,
+                            validators.len()
+                        ))
+                    })
+            })
+            .collect()
     }
 
-    #[test]
-    fn test_bitmap_to_voters_mismatched_lengths() {
-        let validators: Vec<_> = (0..4) // size of 4
-            .into_iter()
-            .map(|_| AccountAddress::random())
-            .collect();
-        let bitmap_too_long = vec![true, true, false, true, true]; // size of 5
-        assert!(ActiveInactiveHeuristic::bitmap_to_voters(&validators, &bitmap_too_long).is_err());
-        let bitmap_too_short = vec![true, true, false];
-        assert!(ActiveInactiveHeuristic::bitmap_to_voters(&validators, &bitmap_too_short).is_err());
+    pub fn history_iter(
+        history: &[NewBlockEvent],
+        epoch: u64,
+        window_size: usize,
+    ) -> impl Iterator<Item = &NewBlockEvent> {
+        let start = if history.len() > window_size {
+            history.len() - window_size
+        } else {
+            0
+        };
+
+        (&history[start..])
+            .iter()
+            .filter(move |&meta| meta.epoch() == epoch)
```

Comment:
we probably should filter this out before, also `move` doesn't seem needed

Reply:
move is needed to transfer ownership for epoch to the closure